swap/src/cli/trace.rs

```diff
@@ -0,0 +1,147 @@
+        let (non_blocking, guard) = tracing_appender::non_blocking(appender);
+        (
+            fmt::layer()
+                .with_ansi(is_terminal)
+                .with_target(false)
+                .with_writer(non_blocking),
+            guard,
+        )
+    };
+
+    match command {
+        Command::BuyXmr { .. } => {
+            let (file_logger, guard) = file_logger(&Uuid::new_v4());
+            let registry = Registry::default().with(file_logger);
+            if debug {
+                set_global_default(registry.with(DebugTerminalPrinter::default()))?;
+            } else {
+                set_global_default(registry.with(InfoTerminalPrinter::default()))?;
+            }
+            Ok(Some(guard))
+        }
+        Command::History => {
+            if debug {
+                set_global_default(Registry::default().with(DebugTerminalPrinter::default()))?;
+            } else {
+                set_global_default(Registry::default().with(InfoTerminalPrinter::default()))?;
+            }
+            Ok(None)
+        }
+        Command::Resume { swap_id, .. }
+        | Command::Cancel { swap_id, .. }
+        | Command::Refund { swap_id, .. } => {
+            let (file_logger, guard) = file_logger(swap_id);
+            let registry = Registry::default().with(file_logger);
+            if debug {
+                set_global_default(registry.with(DebugTerminalPrinter::default()))?;
+            } else {
+                set_global_default(registry.with(InfoTerminalPrinter::default()))?;
+            }
+            Ok(Some(guard))
+        }
+    }
+}
+
+impl<S> Default for DebugTerminalPrinter<S>
+where
+    S: Subscriber + for<'a> LookupSpan<'a>,
+{
+    fn default() -> Self {
+        let is_terminal = atty::is(atty::Stream::Stderr);
+        DebugTerminalPrinter {
+            fmt: fmt::layer()
+                .with_ansi(is_terminal)
+                .with_target(false)
+                .with_timer(ChronoLocal::default()),
+        }
+    }
+}
+
+impl<S> Default for InfoTerminalPrinter<S>
+where
+    S: Subscriber + for<'a> LookupSpan<'a>,
+{
+    fn default() -> Self {
+        let is_terminal = atty::is(atty::Stream::Stderr);
+        InfoTerminalPrinter {
+            fmt: fmt::layer()
+                .with_ansi(is_terminal)
+                .with_target(false)
+                .with_level(false)
+                .without_time(),
+        }
+    }
+}
+
+pub struct TerminalPrinter<S, T> {
+    pub level_filter: LevelFilter,
+    pub fmt: tracing_subscriber::fmt::Layer<
+        S,
+        DefaultFields,
+        Format<tracing_subscriber::fmt::format::Full, T>,
+        fn() -> std::io::Stdout,
+    >,
+}
+
+pub struct DebugTerminalPrinter<S> {
+    pub fmt: tracing_subscriber::fmt::Layer<
+        S,
+        DefaultFields,
+        Format<tracing_subscriber::fmt::format::Full, ChronoLocal>,
+        fn() -> std::io::Stdout,
+    >,
+}
```

Comment:
Usually, you have an easier time in Rust if you make a struct overly generic by saying:

```suggestion
pub struct DebugTerminalPrinter<L> {
    pub inner: L
}
```

And then only specify the trait bounds and/or generic parameters on `impl` blocks whenever necessary.

In this case, this might allow you to just have a generic `TerminalPrinter` with different constructors for different usecases because the difference in the type-signature (whether or not to print time) is outside of the struct definition.